### DISCUSSION TITLE
Broken vertical camera bounds for stages with zoom

### DISCUSSION THREAD

**Discussion by Kasasagi77** [ikemen-engine/Ikemen-GO#1642] [Feature Request] [answered]
### Describe the bug

The latest changes to the camera bounds behavior for stages by @Eiton in #1633 and #1621 are in my opinion a huge problem, because of several reasons:

1. If I tested this correctly, then they break one of the best stage bug-fixes that Ikemen has versus Mugen 1.1: [Skipped zoom-out processing of boundhigh](https://github.com/ikemen-engine/Ikemen-GO/wiki/Stage-features#skipped-zoomout-processing-of-boundhigh) or they break the upper bound in a similay way.<p> 
2. They casually and totally break backwards-compatibility with all zoom stages made specifically for Ikemen by different creators up to this point, and possibly even with zoom-less stages (although I have not tested the second part).<p> 
3. Moreover the desired improvement in Mugen compatibility of these changes is highly debatable - the stages still do not behave like in Mugen 1.1 even if we assume that re-implementing the Mugen 1.1 bug mentioned in point 1 above was the goal. Additionally, I do not think that re-implementing Mugen 1.1 bugs should be the goal here, the previous Ikemen behavior was intentionally better.

---

**Additional thoughts on backwards compatibility.** 

Ikemen had a certain consistent behavior when it comes to vertical stage camera bounds for a long time. In this time creators have made stages or stage variants specifically for Ikemen, and all of these will most likely be broken. This is not a local or specific compatibility breakage. It affects all stages, because all stages have some lower and upper camera bounds.

Moreover since Ikemen fixed upper bound processing in zoom stages ([Skipped zoom-out processing of boundhigh](https://github.com/ikemen-engine/Ikemen-GO/wiki/Stage-features#skipped-zoomout-processing-of-boundhigh)), it was already expected that zoom stages needed to be Ikemen-specific to have a properly behaving upper bound, and this lack of compatibility with Mugen 1.1 was intentional, so increasing the compatibility in the case of vertical camera bounds is, in my opinion, very undesirable.

I won't lie, I have a stake in this: I have created mote than 100 Ikemen-specific stage edits, that will all be broken after these changes, and more, if I understand correctly - I suspect they will not even be fixable if the upper bound is as broken as in Mugen 1.1. Even if these stages are fixable, I will still have to test and fix all of them, and this applies to all creators with Ikemen-specific stages.

This kind of total backwards-compatibility breakage will without a doubt have a very discouraging effect on any creators wanting make content for Ikemen, at least it is extremely discouraging to me.

### To Reproduce

I upload an example of my stage edits for testing:  [graveyrd.zip](https://github.com/ikemen-engine/Ikemen-GO/files/14324642/graveyrd.zip).

This `zip` contain a zooming stage with different/separate `.def` variants, specifically for Mugen 1.1 and for Ikemen Go [db6a8d8f] (before the breaking changes mentioned in this issue).

---

You can use [graveyrd.zip](https://github.com/ikemen-engine/Ikemen-GO/files/14324642/graveyrd.zip) to observe the desired and undesired effects and the total compatibility breakage:

1. Keep the characters close together, for maximum zoom-in, and apart, for maximum zoom-out.<p>
2. Jump high or double-jump so that the camera moves to its upper vertical limit.<p>
3. Run the `*-ikemen.def` variant in Ikemen [db6a8d8f] (before the breaking changes) - note how the background final vertical position is exactly the same, regardless of zoom - this is the desired behavior.<p>
4. Run the `*-mugen.def` variant in Mugen 1.1 - note how the background scrolls differently depending on the current zoom - this is an undesireable bug in Mugen 1.1.<p>
5. Run the `*-ikemen.def` variant in newest Ikemen - see that the stage is totally broken (this shows a severe breaking of backwards-compatibility).<p>
6. Run the `*-mugen.def` variant in newest Ikemen:

    a. see that the lower bound is broken, but it works fine in Mugen 1.1, making any compatibility improvements questionable,

    b. see that the final vertical position of the background is zoom-dependent, re-implementing the bug from Mugen 1.1, and making it even worse (even less of the background is shown than in Mugen 1.1).

### Expected behavior

In my opinion the mentioned changes should simply be reverted, restoring the old behavior.

If this is not possible, then I suggest making the new behavior opt-in only, by using a specific stage parameter. I do not even think that using the generic `ikemenversion` is appropriate here, as I think it is extremely desirable to opt-out of these changes to the vertical camera bounds specifically, as opposed to a whole swath of potential Ikemen features (which using `ikemenversion` would imply).

### Engine Version (or source code date)

develop

### Operating system

Linux

**Comment by K4thos** [comment]
Our main goal is to make sure everything works as mugen 1.1 by default. Stages are one of the least compatible aspect of the engine at the moment and every work that makes us closer to mitigate this problem is very much welcome, so such commits won't be reverted. We might think about adding something like an ikemenversion setting to tweak things and make them better, but keeping things compatible with Mugen is our top priority. Anything extra is nice to have, but is not mandatory, it's up to devs (Eiton in this case) if they want to put in the extra work to add such options.

We're not considering these kind of changes as bugs, so this ticket should be transferred to discussion, I think.

**Comment by potsmugen** [comment]
K4thos beat me to it but here it goes anyway. I agree it should be a discussion.

To start off I think the goal is first and foremost that people don't need to make Ikemen versions of stages to begin with (unless adding the new features). In this regard I think the camera is closer to Mugen right now than in previous versions, albeit still not perfect. I doubt the changes have stopped here, either.

I don't think boundhigh's relation to zoom is necessarily a bug. To be sure I checked Samurai Shodown and that's how it behaves in there (might even have influenced Elecbyte since it was the prominent series with zoom at the time). The same way that changing how it works fixed some stages (by making them *not* like Mugen), it also broke others.

I think Mugen 1.1 parity should be the priority, and then we can think of how to improve it. This parameter in particular sounds like it could use something like "boundhighzoomdelta".

I understand engine updates can be frustrating because I have to deal with them myself as well. But it's all a part of a new engine's growing pains.

EDIT: There's one clever feature that Eiton kept from previous versions that you might have missed: when the stage scrolls past the boundhigh it starts zooming out on its own to show that last piece of it. However this only happens if you use tensionhigh and tensionlow, which are recommended in zooming stages.

**Comment by Eiton** [comment, marked as answer]
Thanks for reporting the issue with your stages. The stage def did not provide the cuthigh value. The engine is not giving a proper default value and the stage fit is not working properly as a result. This will be fixed in #1640.

I understand that you prefer the camera behavior before the fix, the actual bound high value in the old implementation is affected by a lot of parameters (zoom out, boundlow, tension...). I think it is kind of overcomplicated for both stage creators and engine maintainers. (Edit: It doesn't really make sense as well, the camera can move upward even when bound high is a positive value) Maybe we will add back such option in the future but I think Mugen 1.1 behavior would be the default.

**Comment by Kasasagi77** [comment]
Thank you for your responses. 🙏

I see that the Ikemen developers community has a consensus and I am very sad to see it is not in my favor.

I especially counted on the Ikemen features described in the Wiki to not be simply removed in favor on Mugen 1.1 compatibility, but if I understand correctly I can not make any such assumptions until maybe an undetermined time in the future.

> 
> I understand engine updates can be frustrating because I have to deal with them myself as well. But it's all a part of a new engine's growing pains.

I can deal with the frustration and the work required for updating all of my stages.

But if I understand correctly, I will not even be able to fix my stages to match the current behavior with the new Ikemen version. They will be permanently broken in the new Ikemen. And maybe in the years to come there will be some features added that allows me to replicate the current behavior and fix my stages to the current state (and maybe not). Do I understand this correctly?


> Thanks for reporting the issue with your stages. The stage def did not provide the cuthigh value. The engine is not giving a proper default value and the stage fit is not working properly as a result. This will be fixed in #1640.
> 

Will this fix alone allow me to replicate the previous behavior? I suspect not? If I understand correctly then I would need the hypothetical `boundhighzoomdelta` parameter mentioned by @potsmugen to truly replicate the old behavior?

> 
> EDIT: There's one clever feature that Eiton kept from previous versions that you might have missed: when the stage scrolls past the boundhigh it starts zooming out on its own to show that last piece of it. However this only happens if you use tensionhigh and tensionlow, which are recommended in zooming stages.

I very much prefer the old `verticalfollow` behavior. Is there even a way to control scrolling _speed_ with `tensionhigh` and `tensionlow`? Will using `tensionhigh` and `tensionlow` let me replicate the old behavior? I suspect not?

**Comment by potsmugen** [comment]
I usually tell people to make the stages for Mugen 1.1 (unless there's an attached char or something). It's the safest way to make sure they work in future Ikemen.

> I very much prefer the old `verticalfollow` behavior. Is there even a way to control scrolling _speed_ with `tensionhigh` and `tensionlow`? Will using `tensionhigh` and `tensionlow` let me replicate the old behavior? I suspect not?

Older Ikemen versions allowed `verticalfollow` together with `tensionhigh` and `tensionlow`, unlike Mugen. Maybe that could be reconsidered.

**Comment by Kasasagi77** [comment]
I can report that I have tested many of my stages on the newest `develop` (1518892e) and the vertical bounds behavior matches Mugen 1.1 very closely now.

The differences that are left are generally less than one local pixel, reaching one pixel in some edge cases, but I assume this may be good enough?

@Eiton - should I post reports/stages for these edge cases?

**Reply by Eiton** [reply]
The one pixel difference is probably caused by floating point calculation and rounding errors. I think it is good enough.
